src/Bicep.Core/Diagnostics/DisabledDiagnosticsCache.cs

```diff
@@ -0,0 +1,63 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Immutable;
+using System.Linq;
+using Bicep.Core.Extensions;
+using Bicep.Core.Syntax;
+using Bicep.Core.Text;
+
+namespace Bicep.Core.Diagnostics
+{
+    public class DisabledDiagnosticsCache
+    {
+        private readonly ImmutableArray<int> lineStarts;
+        private readonly ProgramSyntax programSyntax;
+        private readonly Lazy<ImmutableDictionary<int, DisableNextLineDirectiveEndPositionAndCodes>> disableNextLineDiagnosticDirectivesCacheLazy;
+
+        public DisabledDiagnosticsCache(ProgramSyntax programSyntax, ImmutableArray<int> lineStarts)
+        {
+            this.programSyntax = programSyntax;
+            this.lineStarts = lineStarts;
+
+            disableNextLineDiagnosticDirectivesCacheLazy = new Lazy<ImmutableDictionary<int, DisableNextLineDirectiveEndPositionAndCodes>>(() => GetDisableNextLineDiagnosticDirectivesCache());
+        }
+
+        public ImmutableDictionary<int, DisableNextLineDirectiveEndPositionAndCodes> DisableNextLineDiagnosticDirectivesCache => disableNextLineDiagnosticDirectivesCacheLazy.Value;
```

Comment:
Rather than exposing this as a property on the class, what about exposing a method instead? e.g.:
```csharp
public DisableNextLineDirectiveEndPositionAndCodes? TryGetDisabledNextLineDirective(int lineNumber)
    => disableNextLineDiagnosticDirectivesCacheLazy.Value.TryGetValue(lineNumber);
```

Benefits are:
* You're able to give the parameter a name (currently it's unclear to the caller what the `int` key corresponds to unless you look at this class's implementation).
* It's less likely that the dictionary will be consumed in a way that'll defeat the laziness - for example, your implementation in BicepFile.cs is doing this:
    ```csharp
    var disabledDiagnosticsCache = new DisabledDiagnosticsCache(ProgramSyntax, lineStarts);
    DisableNextLineDiagnosticDirectivesCache = disabledDiagnosticsCache.DisableNextLineDiagnosticDirectivesCache;
    ```

Reply:
Makes sense. Updated as part of below commit. Thanks!
https://github.com/Azure/bicep/commit/a49bf6e9a4703dc193b93b3e15f745863f9d8ce2